components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -18,89 +22,250 @@ import android.os.ParcelFileDescriptor
 import android.provider.MediaStore
 import androidx.annotation.VisibleForTesting
 import androidx.core.app.NotificationManagerCompat
+import androidx.core.content.FileProvider
 import androidx.core.net.toUri
 import androidx.localbroadcastmanager.content.LocalBroadcastManager
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers.IO
-import kotlinx.coroutines.withContext
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
 import mozilla.components.browser.state.state.content.DownloadState
 import mozilla.components.concept.fetch.Client
 import mozilla.components.concept.fetch.Header
 import mozilla.components.concept.fetch.Headers.Names.CONTENT_LENGTH
 import mozilla.components.concept.fetch.Headers.Names.CONTENT_TYPE
 import mozilla.components.concept.fetch.Headers.Names.REFERRER
+import mozilla.components.concept.fetch.MutableHeaders
 import mozilla.components.concept.fetch.Request
 import mozilla.components.concept.fetch.toMutableHeaders
 import mozilla.components.feature.downloads.ext.addCompletedDownload
 import mozilla.components.feature.downloads.ext.getDownloadExtra
 import mozilla.components.feature.downloads.ext.withResponse
-import mozilla.components.support.base.ids.NotificationIds
+import mozilla.components.support.base.ids.cancel
 import mozilla.components.support.base.ids.notify
 import java.io.File
 import java.io.FileOutputStream
 import java.io.IOException
+import java.io.InputStream
 import java.io.OutputStream
+import kotlin.random.Random
 
 /**
  * Service that performs downloads through a fetch [Client] rather than through the native
  * Android download manager.
  *
  * To use this service, you must create a subclass in your application and it to the manifest.
  */
-abstract class AbstractFetchDownloadService : CoroutineService() {
+@Suppress("TooManyFunctions", "LargeClass")
+abstract class AbstractFetchDownloadService : Service() {
 
     protected abstract val httpClient: Client
     @VisibleForTesting
     internal val broadcastManager by lazy { LocalBroadcastManager.getInstance(this) }
     @VisibleForTesting
     internal val context: Context get() = this
 
-    override fun onCreate() {
-        startForeground(
-            NotificationIds.getIdForTag(context, ONGOING_DOWNLOAD_NOTIFICATION_TAG),
-            DownloadNotification.createOngoingDownloadNotification(context)
-        )
-        super.onCreate()
+    private var listOfDownloadJobs = mutableMapOf<Long, DownloadJobState>()
+
+    private data class DownloadJobState(
+        var job: Job? = null,
```

Comment:
Maybe use immutable data classes and `.copy` instead?

Reply:
While I could see the advantage of this, I don't think it's terribly necessary and would require a somewhat heavy refactor of this code. I feel like it's cleaner to have sections like `onReceive` which directly manipulate the reference in the map which I wouldn't be able to do if I had to use a .copy() (I would instead have to constantly reassign `downloadJobs[downloadId]` to a new copy, which gets a lot messier, at least when I tried to do this).